astropy/units/quantity.py

```diff
@@ -1084,20 +1084,18 @@ def __ilshift__(self, other):
             return NotImplemented
 
         try:
-            factor = self.unit._to(other)
-        except Exception:
-            # Maybe via equivalencies?  Now we do make a temporary copy.
-            try:
-                value = self._to_value(other)
-            except UnitConversionError:
-                return NotImplemented
-
-            self.view(np.ndarray)[...] = value
-
+            factor = self.unit._to(other)  # except Equivalency
+            self_as_array = self.view(np.ndarray)
+            self_as_array *= factor  # except output dtype
+        # The real error is `numpy.core._exceptions._UFuncOutputCastingError`, which
+        # inherits from `TypeError`.
+        except (UnitConversionError, TypeError, AttributeError):
+            # Simple conversion failed. Integer dtype? Conversion via equivalency?
+            # Given other.__rlshift__(self) a chance.
+            return NotImplemented
         else:
-            self.view(np.ndarray)[...] *= factor
+            self._set_unit(other)
```

Comment:
Agreed on strict necessity and clarity. For cyclomatic complexity + type hinting I prefer the one return statement.

```python
try:
   ...
except:
    return ...   # can't avoid this one, but it means there are 2 exits to the function
else:
    return self

return None  # implicit. A good type checker can probably suss that this will never be used, but still...
```

Reply:
Well that's gone.